### ISSUE TITLE
Files are always re-uploaded to S3 storage

### ISSUE THREAD

**Issue by etianen** [etianen/django-s3-storage#43]

**Comment by etianen** [comment]
@dbinetti - I can't reproduce this, I'm afraid. I ran the collectstatic command twice. The first time, everything was uploaded. The second time, nothing was uploaded as all the files were unmodified.

It occurs to me that this may be to do with Heroku's temporary filesystem. Django skips uploading a file if the local path's modification time is less than the remote file's modification time. If Heroku is checking out your project each time, then the timestamps will be newer than the files on S3.

Try running collectstatic from your local filesystem twice.

**Comment by dbinetti** [comment]
Running `collectstatic` from the local filesystem worked.  

```
86 static files copied.
(barberscore-prod)$ django-admin collectstatic

You have requested to collect static files at the destination
location as specified in your settings.

This will overwrite existing files!
Are you sure you want to do this?

Type 'yes' to continue, or 'no' to cancel: yes

0 static files copied, 86 unmodified.
(barberscore-prod)$ 
```

And also confirmed that it doesn't work on Heroku -- but again this appears to be related to the problem I saw in #42   It might all be due to that one single file (which, ironically enough, is no longer being used in DRF.). So not sure you want to spend time debugging for only this situation (Heroku and this one particular file).  Still, there it is...

```
   Successfully installed django-s3-storage-0.11.0
 $ python manage.py collectstatic --noinput
   Traceback (most recent call last):
     File "manage.py", line 11, in <module>
       execute_from_command_line(sys.argv)
     File "/app/.heroku/python/lib/python2.7/site-packages/django/core/management/__init__.py", line 367, in execute_from_command_line
       utility.execute()
     File "/app/.heroku/python/lib/python2.7/site-packages/django/core/management/__init__.py", line 359, in execute
       self.fetch_command(subcommand).run_from_argv(self.argv)
     File "/app/.heroku/python/lib/python2.7/site-packages/django/core/management/base.py", line 294, in run_from_argv
       self.execute(*args, **cmd_options)
     File "/app/.heroku/python/lib/python2.7/site-packages/django/core/management/base.py", line 345, in execute
       output = self.handle(*args, **options)
     File "/app/.heroku/python/lib/python2.7/site-packages/django/contrib/staticfiles/management/commands/collectstatic.py", line 193, in handle
       collected = self.collect()
     File "/app/.heroku/python/lib/python2.7/site-packages/django/contrib/staticfiles/management/commands/collectstatic.py", line 124, in collect
       handler(path, prefixed_path, storage)
     File "/app/.heroku/python/lib/python2.7/site-packages/django/contrib/staticfiles/management/commands/collectstatic.py", line 349, in copy_file
       if not self.delete_file(path, prefixed_path, source_storage):
     File "/app/.heroku/python/lib/python2.7/site-packages/django/contrib/staticfiles/management/commands/collectstatic.py", line 258, in delete_file
       target_last_modified = self.storage.get_modified_time(prefixed_path)
     File "/app/.heroku/python/lib/python2.7/site-packages/django/core/files/storage.py", line 230, in get_modified_time
       dt = self.modified_time(name)
     File "/app/.heroku/python/lib/python2.7/site-packages/django_s3_storage/storage.py", line 322, in modified_time
       return make_naive(self.meta(name)["LastModified"])
     File "/app/.heroku/python/lib/python2.7/site-packages/django_s3_storage/storage.py", line 32, in _do_wrap_errors
       raise IOError("S3Storage error at {!r}: {}".format(name, force_text(ex)))
   IOError: S3Storage error at u'rest_framework/fonts/glyphicons-halflings-regular.woff': An error occurred (404) when calling the HeadObject operation: Not Found
   Error while running '$ python manage.py collectstatic --noinput'.
   See traceback above for details.
```

**Comment by dbinetti** [comment]
But again, it works with 0.9.11.  So what's changed?

**Comment by etianen** [comment]
I'd be really surprised if that one file broke all the unmodified checking
functionality. I'd also be surprised if this worked in 0.9.11, but not now,
since all the logic for this is in collectstatic, and all
django-s3-storages provides is a timestamp...

A few things to try, if you have the patience:

   1. Downgrade to django-s3-storages 0.9.11, and run your heroku build
   again. Does the unmodified checking work now?
   2. Remove django-rest-framework, and run your heroku build again. Does
   the unmodified checking work now? (This is a trickier one to do, since it
   would break your app unless you first fork the app on Heroku)

I realise that you've already got your own thoughts on these checks, but I
feel they need to be explicitly checked again. Otherwise, I'm deeply
confused as to how this behaviour is coming about!

**Comment by dbinetti** [comment]
OK, done.
TL;DR Surprise!

Both worked.  Here are the details (each starting from scratch)


0.9.11 With DRF:
```
remote:      $ python manage.py collectstatic --noinput
remote:        86 static files copied.
```

And then:
```
remote:      $ python manage.py collectstatic --noinput
remote:        1 static file copied, 85 unmodified.
```

With that one static file being the glyphicons file

0.9.11 Without DRF:
```
remote:      $ python manage.py collectstatic --noinput
remote:        68 static files copied.
```

and then
```
remote:      $ python manage.py collectstatic --noinput
remote:        0 static files copied, 68 unmodified.
```

For completeness I tried
0.11.0 without DRF

```
remote:      $ python manage.py collectstatic --noinput
remote:        68 static files copied.
```

And then...
```
remote:      $ python manage.py collectstatic --noinput
remote:        68 static files copied.
```

**Comment by etianen** [comment]
Waaaaaaaaaaaaaaaat????????!!!!!!!!

This makes no goddamn sense. I shall investigate tomorrow.

**Comment by dbinetti** [comment]
lol.  sorry, man!  but think how awesome it will be when you track this one
down...  :-)

**Comment by dbinetti** [comment]
btw, i'm very happy to give you access to the repo and even the heroku
instance.  this is not in full production, so we can mess around quite a bit

**Comment by etianen** [comment]
That's very kind. I've got a hunch to check out, then it may come to that.

Hunch: Given that the only point of contact between collectstatic's change
detection and django-s3-storage is the modified timestamp, then surely
there is something up with the timestamp. Django expects a naive timestamp,
but S3 returns one with a timezone. I suspect that this is causing the
issue.

**Comment by dbinetti** [comment]
If it's relevant, I'm using Django with USE_TZ=True and taking the actual
Timezone from the TZ environment variable.

**Comment by etianen** [comment]
I've updated master with a possible timezone fix. Let me know if it works!

**Comment by dbinetti** [comment]
OK, that worked; the behavior now matches the behavior for v 0.9.11 (meaning, we still have the issue with that pesky .WOFF file, but otherwise existing files are not copied.)

Thanks for tracking this down.  The asset upload is far and away the most expensive component of my deploy workflow.

**Comment by etianen** [comment]
Phew! My hunch was correct. :)

I guess a release is in order. v0.11.1 is currently being built, so should show up on PyPi soon.
